src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/SortableDependencyNode.cs

```diff
@@ -0,0 +1,170 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+using ILCompiler.DependencyAnalysisFramework;
+using Internal.TypeSystem;
+
+namespace ILCompiler.DependencyAnalysis
+{
+    public abstract class SortableDependencyNode : DependencyNodeCore<NodeFactory>
+    {
+#if !SUPPORT_JIT
+        /// <summary>
+        /// Allows grouping of <see cref="ObjectNode"/> instances such that all nodes in a lower phase
+        /// will be ordered before nodes in a later phase.
+        /// </summary>
+        protected internal virtual int Phase => (int)ObjectNodePhase.Unordered;
+
+        /// <summary>
+        /// Gets an identifier that is the same for all instances of this <see cref="ObjectNode"/>
+        /// descendant, but different from the <see cref="ClassCode"/> of any other descendant.
+        /// </summary>
+        /// <remarks>
+        /// This is really just a number, ideally produced by "new Random().Next(int.MinValue, int.MaxValue)".
+        /// If two manage to conflict (which is pretty unlikely), just make a new one...
+        /// </remarks>
+        protected internal abstract int ClassCode { get; }
+
+        // Note to implementers: the type of `other` is actually the same as the type of `this`.
+        protected internal virtual int CompareToImpl(SortableDependencyNode other, CompilerComparer comparer)
+        {
+            throw new NotImplementedException("Multiple nodes of this type are not supported");
+        }
+
+        protected enum ObjectNodePhase
+        {
+            /// <summary>
+            /// Nodes should only be placed in this phase if they have strict output ordering requirements that
+            /// affect compiler correctness. Today that includes native layout tables.
+            /// </summary>
+            Ordered,
+            Unordered
+        }
+
+        protected enum ObjectNodeOrder
```

Comment:
I don't really like how this enum makes it difficult to externally configure the ordering in an assembly outside of ILCompiler.Compiler; however, I don't believe that's worth holding the checkin.

Reply:
We need some way to precisely order the set of nodes that really need it. The enum is mainly for our cognitive reasoning about order. It could easily just be a set of constants carefully chosen in each node's ClassCode property, which is how nodes external to ILCompiler.Compiler would do it. Or did you have an entirely different mechanism in mind?

Reply:
I'm not sure. I'm just expressing discomfort, but not in a useful fashion. So, please check it in.

Reply:
I share your discomfort about this enum :).

I was thinking that maybe instead of specifying the ordering by hand globally, a node could locally declare "I want to be sorted before X, but otherwise I don't care about my order" and have the compiler come up with an ordering by itself. But it might be an overkill.